### PR TITLE
[TR] PIM-4771: fix completeness with local specific

### PR DESCRIPTION
**Pull request by sumbobyboys**
| Q                  | A
| ------------------ | ---
| Migration script?  | N
| Behats             | Y
| Specs & Static     | ?
| Changelog updated  | N

### CODE REVIEW

File: features/product/completeness.feature [hunk 1]
```diff
@@ -274,3 +274,56 @@ Feature: Display the completeness of a product
     When I open the "Completeness" panel
     And I click on the missing "side_view" value for "en_US" locale and "tablet" channel
     Then I should be on the "Media" attribute group
+
+  # @jira  https://akeneo.atlassian.net/browse/PIM-3386
```
Comment (jjanvier) [inline features/product/completeness.feature]: why ?

Comment (sumbobyboys) [inline features/product/completeness.feature]: dropped it, sorry

File: src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php [hunk 2]
```diff
@@ -229,14 +229,18 @@ protected function addRequirementToCompleteness(
...
-            } elseif ($this->productValueCompleteCkecker->supportsValue($productValue)
-                && !$this->productValueCompleteCkecker->isComplete($productValue, $channel, $locale)
-            ) {
-                $missing = true;
-            }
-            if ($missing) {
+
+            $isIncomplete = (null !== $productValue) &&
+                $this->productValueCompleteCkecker->supportsValue($productValue) &&
+                !$this->productValueCompleteCkecker->isComplete($productValue, $channel, $locale);
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php]: 'productValueCompleteCkecker' -> 'productValueCompleteChecker'

File: src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php [hunk 3]
```diff
@@ -229,14 +229,18 @@ protected function addRequirementToCompleteness(
...
+
+            $shouldExistInLocale = !$attribute->isLocaleSpecific() ||
+                (
+                    $attribute->isLocaleSpecific() &&
+                    in_array($locale, $attribute->getLocaleSpecificCodes())
+                );
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php]: specs should be updated, especially because conditions are quite complexes

Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php]: damn, no specs on this file T_T

Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php]: that may explain why there is such complex logic :)

Comment (ClementGautier) [inline src/Pim/Bundle/CatalogBundle/Manager/CompletenessManager.php]: See the comments on the enterprise PR, it could help here.

**Comment by jjanvier** [comment]
GTM

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php [hunk 4]
```diff
@@ -291,28 +291,37 @@ protected function getMainSqlPart()
                     COUNT(distinct v.id)
                     / (
                         SELECT count(*)
-                            FROM pim_catalog_attribute_requirement
-                            WHERE family_id = p.family_id
-                                AND channel_id = c.id
-                                AND required = true
+                        FROM pim_catalog_attribute_requirement r
+                        JOIN pim_catalog_attribute a ON a.id = r.attribute_id
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: as you don't use fields from table `pim_catalog_attribute`, maybe you can remove `JOIN pim_catalog_attribute a ON a.id = r.attribute_id` and do that: 
`LEFT JOIN pim_catalog_attribute_locale al ON al.attribute_id = r.attribute_id`

Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessGenerator.php]: really good catch :+1:

**Pull request merged**